service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java

```diff
@@ -386,7 +386,8 @@ protected final void addTlsKeyCertOptions(final NetServerOptions serverOptions)
 
             final boolean isOpenSslAvailable = OpenSsl.isAvailable();
             final boolean supportsKeyManagerFactory =  OpenSsl.supportsKeyManagerFactory();
-            final boolean useOpenSsl = isOpenSslAvailable && supportsKeyManagerFactory;
+            final boolean useOpenSsl =
+                    getConfig().isNativeTlsRequired() || (isOpenSslAvailable && supportsKeyManagerFactory);
```

Comment:
have you tested what happens when *nativeTlsRequired* is `true` but OpenSSL is not available?

Reply:
Yes, the adapter tries to load OpenSSL anyways and quits then with an `VertxException`: "OpenSSL is not available"

This would be an expected behavior in this case for me...

Stack Trace: 
```
10:01:14.194 [vert.x-eventloop-thread-0] INFO  o.e.h.a.h.v.VertxBasedHttpProtocolAdapter - using OpenSSL [version: null] instead of JDK's default SSL engine
10:01:14.195 [vert.x-eventloop-thread-0] INFO  o.e.h.a.h.v.VertxBasedHttpProtocolAdapter - enabling secure protocol [TLSv1.2]
10:01:14.197 [vert.x-eventloop-thread-0] INFO  o.e.h.a.h.v.VertxBasedHttpProtocolAdapter - enabling client authentication using certificates [org.eclipse.hono.service.auth.TenantApiTrustOptions]
10:01:14.268 [main] ERROR o.e.h.a.h.v.Application$$EnhancerBySpringCGLIB$$8afe9019 - exception occurred during startup, shutting down ...
java.util.concurrent.ExecutionException: io.vertx.core.VertxException: OpenSSL is not available
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.eclipse.hono.service.AbstractApplication.run(AbstractApplication.java:148)
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:723)
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:713)
	at org.springframework.boot.SpringApplication.afterRefresh(SpringApplication.java:703)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:304)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1118)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1107)
	at org.eclipse.hono.adapter.http.vertx.Application.main(Application.java:34)
Caused by: io.vertx.core.VertxException: OpenSSL is not available
	at io.vertx.core.net.impl.SSLHelper.resolveEngineOptions(SSLHelper.java:74)
	at io.vertx.core.net.impl.SSLHelper.<init>(SSLHelper.java:163)
	at io.vertx.core.http.impl.HttpServerImpl.<init>(HttpServerImpl.java:143)
	at io.vertx.core.impl.VertxImpl.createHttpServer(VertxImpl.java:289)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.bindSecureHttpServer(AbstractVertxBasedHttpProtocolAdapter.java:361)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.lambda$doStart$1(AbstractVertxBasedHttpProtocolAdapter.java:180)
	at io.vertx.core.Future.lambda$compose$1(Future.java:265)
	at io.vertx.core.impl.FutureImpl.setHandler(FutureImpl.java:76)
	at io.vertx.core.Future.compose(Future.java:261)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.doStart(AbstractVertxBasedHttpProtocolAdapter.java:168)
	at org.eclipse.hono.service.AbstractProtocolAdapterBase.startInternal(AbstractProtocolAdapterBase.java:259)
	at org.eclipse.hono.service.AbstractServiceBase.start(AbstractServiceBase.java:66)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:483)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:339)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: io.netty.internal.tcnative.SSL
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at io.netty.handler.ssl.OpenSsl.<clinit>(OpenSsl.java:77)
	at org.eclipse.hono.connection.impl.ConnectionFactoryImpl.addTlsTrustOptions(ConnectionFactoryImpl.java:218)
	at org.eclipse.hono.connection.impl.ConnectionFactoryImpl.addOptions(ConnectionFactoryImpl.java:195)
	at org.eclipse.hono.connection.impl.ConnectionFactoryImpl.connect(ConnectionFactoryImpl.java:127)
	at org.eclipse.hono.connection.impl.ConnectionFactoryImpl.connect(ConnectionFactoryImpl.java:105)
	at org.eclipse.hono.client.impl.HonoClientImpl.lambda$connect$5(HonoClientImpl.java:324)
	... 7 common frames omitted
```

Reply:
Hmm, and this is the desired behavior, that the adapter doesn't start at all in this case?
Why shouldn't it be better to start using the JDK's default SSL engine, thus providing a more *graceful degradation*? I understand that you want to verify that the adapter indeed starts up with OpenSSL but you can also find that out by taking a look at the logs during startup ...

Reply:
Well I would say that if you want OpenSSL and it isn't available it should fail. And then you can check the logs. People tend not to look into logs unless something bad happens.

So if you have "bad performance" most likely no one will check the logs for that entry. However if the pod fails to start, then people will take a look.

It would be different if you want a "default" behavior, in which case I would agree that downgrading the a different implementation is a good idea, but not if you explicitly ask for a feature.

Reply:
I agree with ctron. Also from a DevOps perspective I like to have an immediate feedback, that I messed up something with my deployment. If the connector starts it will take a while till this failure is detected ... (E.g. after running the load tests). 

Everyone else who doesn't care about this will anyways not set the `nativeTlsRequired` property.

Reply:
ok, just wanted to be sure you really mean it :-)